flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/logical/PushLimitIntoTableSourceScanRule.scala

```diff
@@ -0,0 +1,127 @@
+  * down. The source need know it should do limit first and do the filter later, it is hard to
+  * implement.
+  */
+class PushLimitIntoTableSourceScanRule extends RelOptRule(
+  operand(classOf[FlinkLogicalSort],
+    operand(classOf[FlinkLogicalTableSourceScan], none)), "PushLimitIntoTableSourceScanRule") {
+
+  override def matches(call: RelOptRuleCall): Boolean = {
+    val sort = call.rel(0).asInstanceOf[Sort]
+    val fetch = sort.fetch
+    val offset = sort.offset
+    // Only push-down the limit whose offset equal zero. Because it is difficult to source based
+    // push to handle the non-zero offset. And the non-zero offset usually appear together with
+    // sort.
+    val onlyLimit = sort.getCollation.getFieldCollations.isEmpty &&
+        (offset == null || RexLiteral.intValue(offset) == 0) &&
+        fetch != null
+
+    var supportPushDown = false
+    if (onlyLimit) {
+      supportPushDown = call.rel(1).asInstanceOf[TableScan]
+          .getTable.unwrap(classOf[TableSourceTable[_]]) match {
+        case table: TableSourceTable[_] =>
+          table.tableSource match {
+            case source: LimitableTableSource[_] => !source.isLimitPushedDown
+            case _ => false
+          }
+        case _ => false
+      }
+    }
+    supportPushDown
+  }
+
+  override def onMatch(call: RelOptRuleCall): Unit = {
+    val sort = call.rel(0).asInstanceOf[Sort]
+    val scan = call.rel(1).asInstanceOf[FlinkLogicalTableSourceScan]
+    val relOptTable = scan.getTable.asInstanceOf[FlinkRelOptTable]
+    val limit = RexLiteral.intValue(sort.fetch)
+    val relBuilder = call.builder()
+    val newRelOptTable = applyLimit(limit, relOptTable, relBuilder)
+    val newScan = scan.copy(scan.getTraitSet, newRelOptTable)
+
+    val newTableSource = newRelOptTable.unwrap(classOf[TableSourceTable[_]]).tableSource
+    val oldTableSource = relOptTable.unwrap(classOf[TableSourceTable[_]]).tableSource
+
+    if (newTableSource.asInstanceOf[LimitableTableSource[_]].isLimitPushedDown
+        && newTableSource.explainSource().equals(oldTableSource.explainSource)) {
+      throw new TableException("Failed to push limit into table source! "
+          + "table source with pushdown capability must override and change "
+          + "explainSource() API to explain the pushdown applied!")
+    }
+
+    call.transformTo(sort.copy(sort.getTraitSet, Collections.singletonList(newScan)))
+  }
+
+  private def applyLimit(
+      limit: Long,
+      relOptTable: FlinkRelOptTable,
+      relBuilder: RelBuilder): FlinkRelOptTable = {
+    val tableSourceTable = relOptTable.unwrap(classOf[TableSourceTable[Any]])
+    val limitedSource = tableSourceTable.tableSource.asInstanceOf[LimitableTableSource[Any]]
+    val newTableSource = limitedSource.applyLimit(limit)
+
+    val statistic = relOptTable.getFlinkStatistic
+    val newRowCount = if (statistic.getRowCount != null) {
+      Math.min(limit, statistic.getRowCount.toLong)
```

Comment:
Do we need to multiply number of partitions?  Because the source may contains more than the limit number of records if it reads multiple splits.

Reply:
It is a logical rule, it is hard to get the partitions. And I think row count is just an estimated value. Both limit number and origin value number is OK to me.